### PR TITLE
WIP Get rid of metastring remnants

### PR DESCRIPTION
**Pull request by hypeJunction**
Needs #11400

### CODE REVIEW

**Comment by jdalsem** [comment]
be advised... upgrading docs have been moved to the appendix

**Comment by hypeJunction** [comment]
This is ready for review. 
One more push with river queries, and we can remove a ton of other redundant sql functions. But I need a break from all this.

**Comment by jdalsem** [comment]
If this is all green it can go in? River queries will be another PR am i right?

**Comment by jdalsem** [comment]
There is a failed unit test in scrutinizer... https://scrutinizer-ci.com/g/Elgg/Elgg/inspections/ef496d01-57ae-46e5-93f0-b0460f2c6dd1 don't know why it reported it as passed though

File: engine/classes/ElggAnnotation.php [hunk 1]
```diff
@@ -75,12 +74,23 @@ public function save() {
 	 * @return bool
 	 */
 	public function delete() {
-		$result = _elgg_delete_metastring_based_object_by_id($this->id, 'annotation');
-		if ($result) {
+		if (!$this->canEdit()) {
```
Comment (jeabakker) [inline engine/classes/ElggAnnotation.php]: shouldn't this be `canDelete`

File: engine/classes/ElggMetadata.php [hunk 2]
```diff
@@ -90,11 +92,24 @@ public function save() {
 	 * @return bool
 	 */
 	public function delete() {
-		$success = _elgg_delete_metastring_based_object_by_id($this->id, 'metadata');
-		if ($success) {
+		if (!$this->canEdit()) {
```
Comment (jeabakker) [inline engine/classes/ElggMetadata.php]: also `canDelete`?

**Review by jeabakker — APPROVED** [review]

**Comment by jeabakker** [comment]
With the additions of `canEdit()` in annotations and metadata, did you check the site as a normal user to make sure everything still works as expected?

**Comment by hypeJunction** [comment]
Oddball, will look later. River will be another PR somewhen

File: engine/classes/ElggAnnotation.php (hunk 1 above)
Comment (hypeJunction) [reply engine/classes/ElggAnnotation.php]: I don't think we have canDelete on annotations

File: engine/classes/ElggMetadata.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/ElggMetadata.php]: Don't think we have that for Metadata

**Comment by hypeJunction** [comment]
can edit was always there, just hidden in the spaghetti code like a mushroom

**Comment by hypeJunction** [comment]
Could you guys restart the scrutinizer build to see if it fixes itself

**Comment by jdalsem** [comment]
no... can't restart a 'succesful' run i think

**Comment by jdalsem** [comment]
just force push the PR so it gets bumped

**Comment by hypeJunction** [comment]
I really don't know why scrutinizer is complaining. Could you guys take a look to see if you see something I don't see?

**Comment by jeabakker** [comment]
> I really don't know why scrutinizer is complaining. Could you guys take a look to see if you see something I don't see?

it says this test fails https://github.com/hypeJunction/Elgg-leaf/blob/426b01faa5e727cb7a7bff72743616ff3582a1db/engine/tests/phpunit/integration/Elgg/Integration/ElggEntityPreloaderIntegrationTest.php#L43

**Comment by hypeJunction** [comment]
I know which test fails, I don't know why. The only reason it fails on scrutinizer and not on Travis is test order. Maybe scrutinizer doesn't have annotations to get at that point.

**Comment by jdalsem** [comment]
i think that is just the simple case... there are no annotations... scrutinizer does not seed, to that may be the reason there is no annotation. Just create the annotations in the test to make sure there is annotation data.

**Comment by hypeJunction** [comment]
All green here

**Comment by jdalsem** [comment]
Nice... On to the next one

**Pull request merged**